api/pkg/transformer/symbol/function/geospatial.go

```diff
@@ -66,6 +74,127 @@ func S2ID(lat interface{}, lon interface{}, level int) (interface{}, error) {
+		if !ok {
+			return nil, errors.New("first point and second point has different format")
+		}
+		if len(secondPoints) != len(firstPoints) {
+			return nil, errors.New("both first point and second point arrays must have the same length")
+		}
+		var values []interface{}
+		for idx, point1 := range firstPoints {
+			point2 := secondPoints[idx]
+			values = append(values, calculatePolarAngle(point1, point2))
+		}
+		return values, nil
+
+	case *LatLong:
+		firstPoint := firstPoint.(*LatLong)
+		secondPoint, ok := secondPoint.(*LatLong)
+		if !ok {
+			return nil, errors.New("first point and second point has different format")
+		}
+		value := calculatePolarAngle(firstPoint, secondPoint)
+		return value, nil
+
+	default:
+		return nil, fmt.Errorf("unknown type: %T", firstPoint)
+	}
+}
+
+// Haversine distance calculate distance between to points (lat, lon) in km
+// lat1, lon1, lat2, lon2 should be float64 or []float64 or any value that can be converted to those types (float64 or []float64)
+func HaversineDistance(lat1 interface{}, lon1 interface{}, lat2 interface{}, lon2 interface{}) (interface{}, error) {
+	firstPoint, err := extractLatLong(lat1, lon1)
+	if err != nil {
+		return nil, err
+	}
+
+	secondPoint, err := extractLatLong(lat2, lon2)
+	if err != nil {
+		return nil, err
+	}
+
+	switch firstPoint.(type) {
+	case []*LatLong:
+		firstPoints := firstPoint.([]*LatLong)
+		secondPoints, ok := secondPoint.([]*LatLong)
+		if !ok {
+			return nil, errors.New("first point and second point has different format")
+		}
+		if len(secondPoints) != len(firstPoints) {
+			return nil, errors.New("both first point and second point arrays must have the same length")
+		}
+		var values []interface{}
+		for idx, point1 := range firstPoints {
+			point2 := secondPoints[idx]
+			values = append(values, calculateDistance(point1, point2))
+		}
+		return values, nil
+
+	case *LatLong:
+		firstPoint := firstPoint.(*LatLong)
+		secondPoint, ok := secondPoint.(*LatLong)
+		if !ok {
+			return nil, errors.New("first point and second point has different format")
+		}
+		value := calculateDistance(firstPoint, secondPoint)
+		return value, nil
+
+	default:
+		return nil, fmt.Errorf("unknown type: %T", firstPoint)
+	}
+}
+
+func degreesToRadians(d float64) float64 {
+	return d * math.Pi / 180
+}
+
+func calculateDistance(firstPoint *LatLong, secondPoint *LatLong) float64 {
+	lat1 := degreesToRadians(firstPoint.lat)
+	lon1 := degreesToRadians(firstPoint.long)
+	lat2 := degreesToRadians(secondPoint.lat)
+	lon2 := degreesToRadians(secondPoint.long)
+
+	diffLat := lat2 - lat1
+	diffLon := lon2 - lon1
+
+	a := math.Pow(math.Sin(diffLat/2), 2) + math.Cos(lat1)*math.Cos(lat2)*
+		math.Pow(math.Sin(diffLon/2), 2)
+
+	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
+	return c * earthRaidusKm
+}
+
+func calculatePolarAngle(firstPoint *LatLong, secondPoint *LatLong) float64 {
+	distance := calculateDistance(firstPoint, secondPoint)
+	if distance < minimumDistance {
```

Comment:
I think the unit is incorrect here, the `minimumDistance` is in meter whereas `calculateDistance` is in km

Reply:
And can you add testcase for this